### PR TITLE
Re-add support for Dota 2 (on Source 2, and x64)

### PR DESCRIPTION
**Pull request by psychonic**
Before this gets any messier (after over 1.5yrs of churn), I'd like to get it upstream.

It's pretty-well tested on Dota, but only basic functionality confirmed on a couple of other games. There shouldn't be much impact to others, since most of it is just added special casing.

This also updates the ambuild files to the newer API for the x64 support.

I still need to do a little more review myself, but am at the point where I'm ready to also have others take a look.

### CODE REVIEW

File: core/metamod.cpp [hunk 1]
```diff
@@ -469,11 +514,26 @@ mm_InitializeGlobals(CreateInterfaceFn engineFactory,
...
+
+static void
+mm_HandleGameInit()
 {
 	if (is_game_init)
-		return true;
+		return;
+
+#if SOURCE_ENGINE == SE_DOTA
+	Msg("MMS: GameInit\n");
```
Comment (KyleSanderson) [inline core/metamod.cpp]: debug printout?

File: core/metamod.cpp [hunk 2]
```diff
@@ -486,24 +546,15 @@ Handler_GameInit()
...
-
-	g_SourceHook.CompleteShutdown();
 }
 
 static void
-Handler_LevelShutdown(void)
+mm_HandleLevelShutdown()
 {
+#if SOURCE_ENGINE == SE_DOTA
+	Msg("MMS: LevelShutdown\n");
```
Comment (KyleSanderson) [inline core/metamod.cpp]: ditto on debug?

File: core/metamod.cpp [hunk 3]
```diff
@@ -534,6 +585,96 @@ Handler_LevelShutdown(void)
...
+static void
+mm_HandleLevelInit(char const *pMapName,
+char const *pMapEntities,
+char const *pOldLevel,
+char const *pLandmarkName,
+bool loadGame,
+bool background)
+{
+#if SOURCE_ENGINE == SE_DOTA
+	Msg("MMS: LevelInit\n");
```
Comment (KyleSanderson) [inline core/metamod.cpp]: debug printout

File: core/metamod_oslink.h [hunk 4]
```diff
@@ -100,10 +100,6 @@ bool GetFileOfAddress(void *pAddr, char *buffer, size_t maxlength);
 	typedef unsigned __int32	uint32_t;
 #elif defined __GNUC__
 #include <stdint.h>
-#if !__GLIBC_HAVE_LONG_LONG
```
Comment (KyleSanderson) [inline core/metamod_oslink.h]: Is this intended? Could a really old glibc/gcc rely on this for an ep1 game?

File: core/provider/provider_ep2.cpp [hunk 5]
```diff
@@ -38,6 +38,18 @@
 #include "metamod_console.h"
 #include <filesystem.h>
 #include "metamod.h"
+#include <tier1/KeyValues.h>
```
Comment (KyleSanderson) [inline core/provider/provider_ep2.cpp]: I'm not sure we linked tier1 here previously? This brings in utlvector and ilk.

File: loader/loader.cpp [hunk 6]
```diff
@@ -89,7 +89,7 @@ static const char *backend_names[] =
 	"2.blade",
 	"2.insurgency",
 	"2.contagion",
-	"2.bms",
```
Comment (KyleSanderson) [inline loader/loader.cpp]: unless if this is triggering a detection, might as well just leave it in.

File: loader/gamedll.cpp [hunk 7]
```diff
@@ -131,12 +165,14 @@ mm_DetectGameInformation()
 			lptr = cur_path;
 		}
 
+		const char *pRelPath = is_source2 ? "../../" : "";
```
Comment (KyleSanderson) [inline loader/gamedll.cpp]: mm_ResolvePath might be helpful here.

File: loader/gamedll.cpp [hunk 8]
```diff
@@ -393,33 +646,169 @@ mm_PatchDllShutdown()
...
+		}
+
+		if (ptr != NULL)
+		{
+			mm_FreeCachedLibraries();
+			gamedll_lib = lib;
+			config_iface = (ISource2ServerConfig *) ptr;
+			gamedll_qvi = qvi;
+
+			mm_PatchConnect(true);
```
Comment (KyleSanderson) [inline loader/gamedll.cpp]: I'm not saying this is wrong, this is pretty required; but we're making a "new" delineation here that MM:S is actively patching the game, enabling clients to connect and the server to even start as a dedicated server. More of a "okay this is changing with MM:S" instead of a "hey whoa there" as things like this have traditionally lived in a plugin to the best of my knowledge.

File: core/metamod.cpp (hunk 1 above)
Comment (psychonic) [inline core/metamod.cpp]: These printouts are intentional, but can be relegated to a DevMsg. They're in the same format of existing printouts in the S2 console, of debug ILK and otherwise.

File: core/metamod.cpp (hunk 2 above)
Comment (psychonic) [inline core/metamod.cpp]: ditto on debug

File: core/metamod.cpp (hunk 3 above)
Comment (psychonic) [inline core/metamod.cpp]: ditto

File: core/metamod_oslink.h (hunk 4 above)
Comment (psychonic) [inline core/metamod_oslink.h]: I don't remember why this change was made, but it should be safe to require modern toolchains. I think we already do anyway for recentish FastDelegate changes (variadic templates).

File: core/provider/provider_ep2.cpp (hunk 5 above)
Comment (psychonic) [inline core/provider/provider_ep2.cpp]: We do already use tier1 for reading VDFs.

File: loader/gamedll.cpp (hunk 8 above)
Comment (psychonic) [inline loader/gamedll.cpp]: I considered that. The problem is that the process would exit out before the point that we get to plugin loading.

One minor correction - I think we're still not directly allowing clients to join, at least not in Dota. I believe they still have to pass another check that would need to be bypassed. This change only allows to server to not exit on game init.

We are actively patching, intentionally and out of necessity.

Comment (psychonic) [inline loader/gamedll.cpp]: Right after replying, I realize that I half misunderstood. The name PatchConnect comes from patching the Connect function on the IAppSystem impl, just like we do for DllInit on the earlier engines. It has nothing to do with clients connecting.

We are however patching ISource2ServerConfig::AllowDedicatedServers, as it returns false on Windows at start, unless running in the Steam Beta or Dev universe.

File: loader/loader.cpp (hunk 6 above)
Comment (psychonic) [inline loader/loader.cpp]: This looks like a merge gone wrong. Looking into it...

File: loader/gamedll.cpp (hunk 7 above)
Comment (psychonic) [inline loader/gamedll.cpp]: Looking into it...

Comment (psychonic) [inline loader/gamedll.cpp]: I don't think that's what we want here since we're dealing with relative paths for comparison.

File: core/AMBuilder [hunk 9]
```diff
@@ -16,13 +16,16 @@ for sdk_name in MMS.sdks:
...
-    'sourcehook/sourcehook_hookmangen.cpp',
     'sourcehook/sourcehook_impl_chookidman.cpp',
     'sourcehook/sourcehook_impl_chookmaninfo.cpp',
     'sourcehook/sourcehook_impl_cproto.cpp',
     'sourcehook/sourcehook_impl_cvfnptr.cpp',
     'gamedll_bridge.cpp',
     'vsp_bridge.cpp'
   ]
+
+  # Source2 hack. TODO: check this more deterministically, "are we doing an x64 build?"
```
Comment (KyleSanderson) [inline core/AMBuilder]: Might be worthwhile to builder.target.arch instead here.

Comment (psychonic) [inline core/AMBuilder]: Good call. Fixed this now.
